rl4co/envs/routing/pdp/env.py

```diff
@@ -209,3 +209,317 @@ def check_solution_validity(self, td, actions):
+        
+        argsort = rec.argsort()
+        
+        pre_pairsecond = argsort.gather(1, pair_index + gs // 2)
+        post_pairsecond = rec.gather(1, pair_index + gs // 2)
+        
+        rec.scatter_(1,pre_pairsecond,post_pairsecond)
+        
+        # fix connection for pairing node
+        post_second = rec.gather(1,second)
+        rec.scatter_(1,second, pair_index + gs // 2)
+        rec.scatter_(1,pair_index + gs // 2, post_second)
+        
+        post_first = rec.gather(1,first)
+        rec.scatter_(1,first, pair_index)
+        rec.scatter_(1,pair_index, post_first)        
+        
+        return rec
+
+    def _make_spec(self, generator: PDPGenerator):
+        """Make the observation and action specs from the parameters."""
+        self.observation_spec = CompositeSpec(
+            locs=BoundedTensorSpec(
+                low=generator.min_loc,
+                high=generator.max_loc,
+                shape=(generator.num_loc + 1, 2),
+                dtype=torch.float32,
+            ),
+            cost_current=UnboundedContinuousTensorSpec(
+                shape=(1),
+                dtype=torch.float32,
+            ),
+            cost_bsf=UnboundedContinuousTensorSpec(
+                shape=(1),
+                dtype=torch.float32,
+            ),
+            rec_current=UnboundedDiscreteTensorSpec(
+                shape=(self.generator.num_loc + 1),
+                dtype=torch.int64,
+            ),
+            rec_best=UnboundedDiscreteTensorSpec(
+                shape=(self.generator.num_loc + 1),
+                dtype=torch.int64,
+            ),
+            visited_time=UnboundedDiscreteTensorSpec(
+                shape=(self.generator.num_loc + 1, self.generator.num_loc + 1),
+                dtype=torch.int64,
+            ),
+            action_record=UnboundedDiscreteTensorSpec(
+                shape=(self.generator.num_loc + 1, self.generator.num_loc + 1),
+                dtype=torch.int64,
+            ),
+            i=UnboundedDiscreteTensorSpec(
+                shape=(1),
+                dtype=torch.int64,
+            ),
+            action_mask=UnboundedDiscreteTensorSpec(
+                shape=(self.generator.num_loc + 1, self.generator.num_loc + 1),
+                dtype=torch.bool,
+            ),
+            shape=(),
+        )
+        self.action_spec = BoundedTensorSpec(
+            shape=(3,),
+            dtype=torch.int64,
+            low=0,
+            high=self.generator.num_loc + 1,
+        )
+        self.reward_spec = UnboundedContinuousTensorSpec(shape=(1,))
+        self.done_spec = UnboundedDiscreteTensorSpec(shape=(1,), dtype=torch.bool)
+
+    @staticmethod
+    def _get_reward(td, actions) -> TensorDict:
+        # This function is not used for improvement tasks since the reward is computed per step
+        assert False
+
+    def check_solution_validity(self, td, actions):
+        pass
+        #Todo
```

Comment:
Yeah, the solution will always be valid, and I have added the codes here. It can be called by the agent to check the validity of the best found solution so far